### PR TITLE
Fixes for `decorators` property deprecations

### LINKED ISSUE #50259: Deprecated decorator factory APIs throw "Invalid arguments" runtime error
# Bug Report

The following factory methods, that got deprecated in 4.8, throw an "Invalid arguments" type error at runtime, if the `body` parameter is not `undefined`:

```
t.factory.createConstructorDeclaration(decorators, modifiers, parameters, body);
t.factory.updateConstructorDeclaration(ctor, decorators, modifiers, parameters, body);
```

### 🔎 Search Terms

Invalid arguments, createConstructorDeclaration, updateConstructorDeclaration

### 🕗 Version & Regression Information

- This changed between versions 4.7.4 and 4.8.0-dev.20220809

### ⏯ Playground Link

I don't think this can be reproduced in the playground.

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.0-beta#code/MYewdgzgLgBFEwLwwE4FMCOBXAlugFAORQCeADmhMCjmVIQJQDcAUC6JLAEYA2IwAayRwIAOgBmAQ2BQQKEqOppJUNACE+g-AG0AuszYB6QzADucgRHbhoMGXOHwJ02fMXoVaAMI2oKLPYoACJowDySKCo44PgsMDCGAFQwALYgACY44jhoKAiJJlhg6WjZYGjpADRxCclkEZIpaKp5MAUwetXxSTBcGSRtJrz8AtUGLMZmFlYctgD6WGTpnulergByaKYAgmQ4jmJSgQqLy6o+nP6BIWENUNFgsfGBXbWpGVk5re2BommZ2VyEFePXqkSaLXyJl+YMazSBIOSfXSAx+rlEyJIY1YExMUAAFigQKYEAAVchoACiKCJKHwACIAJJgABukh4OHSMAiAHMsE0wPB6QZZrA5oEAPI8dK7fbIJxHVwKJSeC7QK6uG7hSL3GI1HolUA6uRQmBFEplCqI94Ar6m82lHDlKr6uoNCFAwYdXTWzFe4aCbFGPGE4lkinU2kM5lsjlc3n8tCCiDC1iimALJYrNZyEJkdDAFayg7OY6iU6q3wa3OhbVRPXPVzWw1yFQmr2-FvGvLW-6fT1ouR-D6AvIAfgkOB4LXwThwEAAsiOvgxrbCPd9oej1-Ce67ev0O+jMdigA)

### 💻 Code

To reproduce, you can:

```sh
$ mkdir repro
$ npm init --yes
$ npm install typescript@4.8.0-dev.20220809
```

Add the following code in a file `index.js`. Then run `node index.js`.

```ts
const ts = require('typescript');

const block = ts.factory.createBlock([]);

// works
const ctor = ts.factory.createConstructorDeclaration(
  /* modifiers */ undefined,
  /* parameters */ [],
  /* body */ block,
);

// works
const _updatedCtorNewApi = ts.factory.updateConstructorDeclaration(
  ctor,
  /* modifiers */ ctor.modifiers,
  /* parameters */ ctor.parameters,
  /* body */ ctor.body,
);

// throws TypeError("Invalid arguments");
const _ctorOldApi = ts.factory.createConstructorDeclaration(
  /* decorators */ undefined,
  /* modifiers */ undefined,
  /* parameters */ [],
  /* body */ block,
);

// throws TypeError("Invalid arguments");
const _updatedCtorDeprecatedApi = ts.factory.updateConstructorDeclaration(
  ctor,
  /* decorators */ ctor.decorators,
  /* modifiers */ ctor.modifiers?.filter(ts.isModifier),
  /* parameters */ ctor.parameters,
  /* body */ ctor.body,
);
```

### 🙁 Actual behavior

The deprecated functions createConstructorDeclaration and updateConstructorDeclaration throw an error at runtime. This makes it hard to write code, which works with both TypeScript 4.7 and 4.8.

### 🙂 Expected behavior

No error 🙂 

### 🕵️ Cause

If I understand the deprecation function overload logic correctly, the issue is here:

https://github.com/microsoft/TypeScript/blob/3c3909b8d0a3d81cf320a84dac373eac19de3a7a/src/deprecatedCompat/4.8/mergeDecoratorsAndModifiers.ts#L538

https://github.com/microsoft/TypeScript/blob/3c3909b8d0a3d81cf320a84dac373eac19de3a7a/src/deprecatedCompat/4.8/mergeDecoratorsAndModifiers.ts#L566

This should use `isBlock(body)` (without the negation).

### PR DESCRIPTION
**Pull request by rbuckton**
This fixes two issues that occur in TS 4.8 related to the deprecation of the `decorators` property on `Node`:

First, the deprecated `decorators` property used a type that made it difficult to find runtime breaks since use of `decorators` would silently succeed rather than fail to compile. To address this, I've changed the typeof `decorators` from `NodeArray<Decorator> | undefined` to `undefined` to make these breaks more obvious. As a result of this change, I renamed the `decorators` property to `_decorators` internally for cases where we track invalid grammar nodes (such as decorators parsed on a constructor, etc.), otherwise I would have run into a conflict with the definition of the types.

Second, the `createConstructorDeclaration` and `updateConstructorDeclaration` functions of `NodeFactory` incorrectly rejected a valid value for `body` when using the deprecated overloads due to an incorrect negation of a test.

Fixes #50259

### CODE REVIEW

File: tests/baselines/reference/api/typescript.d.ts [hunk 1]
```diff
@@ -7912,9 +7912,27 @@ declare namespace ts {
...
+         * decorators of a `Node` that can legally have decorators, use the `ts.getDecorators()` function.
+         *
+         * For example:
+         * ```ts
+         * const decorators = ts.canHaveDecorators(node) ? ts.getDecorators(node) : undefined;
+         * ```
+         */
+        readonly decorators: never;
+        /**
+         * @deprecated `modifiers` has been removed from `Node` and moved to the specific `Node` subtypes that support them.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/api/typescript.d.ts:7927]: are `modifiers` actually removed? or just more specific in their type on sub-nodes?

Comment (rbuckton) [reply tests/baselines/reference/api/typescript.d.ts:7927]: It's intended that `modifiers` is not a property of every `Node`, but rather is a property of specific `Node` sub-types. i.e., you shouldn't just blindly read from `node.modifiers`.

File: src/deprecatedCompat/4.8/mergeDecoratorsAndModifiers.ts [hunk 2]
```diff
@@ -5,10 +5,28 @@
 //     - error: 5.0
 namespace ts {
     export interface Node {
-        /** @deprecated `decorators` has been merged with `modifiers` on the declarations that support decorators. */
-        readonly decorators?: NodeArray<Decorator> | undefined;
+        /**
+         * @deprecated `decorators` has been merged with `modifiers` on the declarations that support decorators.
+         * To test whether a `Node` can legally have decorators, use the `ts.canHaveDecorators()` function. To read the
+         * decorators of a `Node` that can legally have decorators, use the `ts.getDecorators()` function.
```
Comment (sandersn) [inline src/deprecatedCompat/4.8/mergeDecoratorsAndModifiers.ts:11]: I got very confused by the last sentence. I'd recommend rewriting it:

```suggestion
         * To test whether a `Node` can legally have decorators, use the `ts.canHaveDecorators()` function.
         * Use `ts.getDecorators()` to get decorators from `Node`s that have them.
```

**Review by sandersn — APPROVED** [review]
Looks good although I got confused by the wording of the last sentence of `decorators`' jsdoc.

**Comment by DanielRosenwasser** [comment]
@typescript-bot cherry-pick this to release-4.8

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by rbuckton** [comment]
@DanielRosenwasser, we may need to re-create or update the cherry-pick PR.

**Comment by DanielRosenwasser** [comment]
@typescript-bot cherry-pick this to release-4.8

**Pull request merged**